src/nexmark/queries/q19.rs

```diff
@@ -0,0 +1,289 @@
+use super::NexmarkStream;
+use crate::{
+    nexmark::model::{Bid, Event},
+    operator::{FilterMap, Fold},
+    Circuit, OrdZSet, Stream,
+};
+use std::collections::VecDeque;
+
+///
+/// Query 19: Auction TOP-10 Price (Not in original suite)
+///
+/// What's the top price 10 bids of an auction?
+/// Illustrates a TOP-N query.
+///
+/// ```sql
+/// CREATE TABLE discard_sink (
+///     auction  BIGINT,
+///     bidder  BIGINT,
+///     price  BIGINT,
+///     channel  VARCHAR,
+///     url  VARCHAR,
+///     dateTime  TIMESTAMP(3),
+///     extra  VARCHAR,
+///     rank_number  BIGINT
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT * FROM
+/// (SELECT *, ROW_NUMBER() OVER (PARTITION BY auction ORDER BY price DESC) AS rank_number FROM bid)
+/// WHERE rank_number <= 10;
+/// ```
+
+type Q19Stream = Stream<Circuit<()>, OrdZSet<Bid, isize>>;
+
+const TOP_BIDS: usize = 10;
+
+pub fn q19(input: NexmarkStream) -> Q19Stream {
+    let bids_by_auction = input.flat_map_index(|event| match event {
+        Event::Bid(b) => Some((b.auction, b.clone())),
+        _ => None,
+    });
+
+    bids_by_auction
+        .aggregate::<(), _>(Fold::new(
+            VecDeque::with_capacity(TOP_BIDS),
+            |top: &mut VecDeque<Bid>, val: &Bid, _w| {
+                if top.len() >= TOP_BIDS {
+                    top.pop_front();
+                }
+                top.push_back(val.clone());
+            },
+        ))
+        .flat_map(|(_, vec)| -> VecDeque<Bid> { (*vec).clone() })
```

Comment:
Not sure yet if there's a better way, but because the Fold needs an `&mut VecDeque<Bid>`, the vec here is also a reference, which means that without dereferencing the vec (and then requiring a clone :/ ), the output type is also a referenc:

```diff
diff --git a/src/nexmark/queries/q19.rs b/src/nexmark/queries/q19.rs
index 729f6f0..3f45aa5 100644
--- a/src/nexmark/queries/q19.rs
+++ b/src/nexmark/queries/q19.rs
@@ -52,7 +52,7 @@ pub fn q19(input: NexmarkStream) -> Q19Stream {
                 top.push_back(val.clone());
             },
         ))
-        .flat_map(|(_, vec)| -> VecDeque<Bid> { (*vec).clone() })
+        .flat_map(|(_, vec)| vec)
 }
 
 #[cfg(test)]
```

```
error[E0308]: mismatched types
  --> src/nexmark/queries/q19.rs:45:5
   |
39 |   pub fn q19(input: NexmarkStream) -> Q19Stream {
   |                                       --------- expected `Stream<circuit_builder::Circuit<()>, zset_batch::OrdZSet<nexmark::model::Bid, isize>>` because of return type
...
45 | /     bids_by_auction
46 | |         .aggregate::<(), _>(Fold::new(
47 | |             VecDeque::with_capacity(TOP_BIDS),
48 | |             |top: &mut VecDeque<Bid>, val: &Bid, _w| {
...  |
54 | |         ))
55 | |         .flat_map(|(_, vec)| vec)
   | |_________________________________^ expected struct `nexmark::model::Bid`, found `&nexmark::model::Bid`
   |
   = note: expected struct `Stream<_, zset_batch::OrdZSet<nexmark::model::Bid, _>>`
              found struct `Stream<_, zset_batch::OrdZSet<&nexmark::model::Bid, _>>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `dbsp` due to previous error
```